File: test/keyw.lua
local ffi = require('ffi')
local keyw = ffi.load('test/keyw.so')

ffi.cdef [[

typedef enum {
    TEOF,       // End of file
    TWord,      // any
    TIONumber,  // Integer positive number delimited by '<' or '>'
    TNewLine,   // \n
    TAnd,       // &
    TOr,        // |
    TSemi,      // ;
    TAndIf,     // &&
    TOrIf,      // ||
    TDSemi,     // ;;
    TLess,      // <
    TGreat,     // >
    TDLess,     // <<
    TDGreat,    // >>
    TLessAnd,   // <&
    TGreatAnd,  // >&
    TLessGreat, // <>
    TDLessDash, // <<-
    TLobber,    // >|
    TIf,        // if
    TThen,      // then
    TElse,      // else
    TElif,      // elif
    TFi,        // fi
    TDo,        // do
    TDone,      // done
    TCase,      // case
    TEsac,      // esac
    TWhile,     // while
    TUntil,     // until
    TFor,       // for
} TokenType;

TokenType keyw_gettype(const char *);

]]

TokenType = {
	TWord = 1,
	TIf = 19,
	TThen = 20,
	TElse = 21,
	TElif = 22,
	TFi = 23,
	TDo = 24,
	TDone = 25,
	TCase = 26,
	TEsac = 27,
	TWhile = 28,
	TUntil = 29,
	TFor = 30,
}

local tests = {
	{keyword = "if", want = TokenType.TIf},
	{keyword = "then", want = TokenType.TThen},
	{keyword = "else", want = TokenType.TElse},
	{keyword = "elif", want = TokenType.TElif},
	{keyword = "fi", want = TokenType.TFi},
	{keyword = "do", want = TokenType.TDo},
	{keyword = "done", want = TokenType.TDone},
	{keyword = "case", want = TokenType.TCase},
	{keyword = "esac", want = TokenType.TEsac},
	{keyword = "while", want = TokenType.TWhile},
	{keyword = "until", want = TokenType.TUntil},
	{keyword = "for", want = TokenType.TFor},
	{keyword = "ifforuntil", want = TokenType.TWord},
}

print '\tkeyw test: keywords'
for k, t in pairs(tests) do
	local got = keyw.keyw_gettype(t.keyword)

	if got ~= t.want then
		print(string.format("\tkeyword.type test at k=%d: got=%s, \z
			want=%s", k, got, t.want))
	end
end

File: src/keyw.c
#include <stdbool.h>
#include "lex.h"

#define LENGTH 12

static char *keywords[LENGTH] = {
    "if",
    "then",
    "else",
    "elif",
    "fi",
    "do",
    "done",
    "case",
    "esac",
    "until",
    "while",
    "for",
};

static int keywordtypes[LENGTH] = {
    TIf,
    TThen,
    TElse,
    TElif,
    TFi,
    TDo,
    TDone,
    TCase,
    TEsac,
    TUntil,
    TWhile,
    TFor,
};

TokenType keyw_gettype(const char *);
static bool streq(const char *, const char *);

// ---------------------------------------------------------------------------

// keyw_typeof returns the TokenType of the keyword provided.  If the
// provided keyword is, actually, not a keyword, keyw_gettype returns the type
// TWord.
TokenType keyw_typeof(const char *keyword)
{
    for (int i = 0; i < LENGTH; i++) {
	if (streq(keyword, keywords[i])) {
	    return keywordtypes[i];
	}
    }
    return TWord;
}

// streq checks whether s1 and s2 are equal.
static bool streq(const char *s1, const char *s2)
{
    while (*s1) {
	s1++;
	s2++;

	if (*s1 != *s2) {
	    return false;
	}
    }
    return true;
}
